api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -433,16 +449,51 @@ private User initializeUserIfNeeded() {
   }
 
   @Override
-  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest){
+  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest) {
     verifyInvitationKey(invitationVerificationRequest.getInvitationKey());
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
 
-  private void verifyInvitationKey(String invitationKey){
-    if(invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
+  private void verifyInvitationKey(String invitationKey) {
+    if (invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
       throw new BadRequestException(
-        "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+          "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+    }
+  }
+
+  /*
+   * This un-authed API method is limited such that we only allow contact email updates before the user has signed in
+   * with the newly created gsuite account. Once the user has logged in, they can change their contact email through
+   * the normal profile update process.
+   */
+  @Override
+  public ResponseEntity<Void> updateContactEmail(UpdateContactEmailRequest updateContactEmailRequest) {
+    User user = userDao.findUserByEmail(updateContactEmailRequest.getUsername());
+    if (user.getFirstSignInTime() != null) {
```

Comment:
Thinking more on this, the ability to change contact email does also introduce some holes in terms of email verification. Happy to go over the various cases offline, but in general I think it would be more secure to do the following:

- If possible, check against gsuite somehow to check for whether they've already logged in, e.g. setup a new password, setup recovery options, instead of checking for workbench sign in.
- Move the password reset (and perhaps the resend of welcome email) onto the server side to guarantee that contact email and pw reset are happening together (secure ordering would be reset pw -> update contact email).

Reply:
I'm not following how this would lock them out. At this point they've yet to login to the account and currently the client code just calls updateContactEmail() -> resendWelcomeEmail(), effectively resulting in the same behavior I'm proposing. My suggestion was just to do these two things on the server rather than the client.

example:
- I own alice@gmail.com but not bob@gmail.com
- create account with contact of alice@gmail.com; password is now foo
- call updateContactEmail API directly with bob@gmail.com; password still foo
- Login using the welcome email password
- My user now has a verified email of bob@gmail.com

Reply:
I better understand what you're proposing CH, I did misunderstand before. Making a combination call to do both will definitely work. My lock-out comment would only happen if we added the password reset to this method - my misunderstanding there.

In the example flow, step 3 would fail because the user had not yet signed in. This specific version of updating the contact email has the same restriction as sending the welcome email.